atc/metric/emitter/newrelic.go

```diff
@@ -1,41 +1,74 @@
 package emitter
 
 import (
+	"bufio"
 	"bytes"
+	"compress/gzip"
 	"encoding/json"
 	"fmt"
+	"github.com/concourse/flag"
+	"io"
+	"io/ioutil"
 	"net/http"
 	"strings"
+	"sync"
 	"time"
 
 	"code.cloudfoundry.org/lager"
 	"github.com/concourse/concourse/atc/metric"
 	"github.com/pkg/errors"
 )
 
+const (
+	NEWRELIC_PAYLOAD_MAX_SIZE            = 1024 * 1024
+	NEWRELIC_FLUSH_CHECK_INTERVAL_IN_SEC = 30
+)
+
 type (
 	stats struct {
 		created interface{}
 		deleted interface{}
 	}
 
 	NewRelicEmitter struct {
-		client     *http.Client
-		url        string
-		apikey     string
-		prefix     string
-		containers *stats
-		volumes    *stats
+		prefix       string
+		containers   *stats
+		volumes      *stats
+		batchEmitter *batchEmitter
 	}
 
 	NewRelicConfig struct {
-		AccountID     string `long:"newrelic-account-id" description:"New Relic Account ID"`
-		APIKey        string `long:"newrelic-api-key" description:"New Relic Insights API Key"`
-		ServicePrefix string `long:"newrelic-service-prefix" default:"" description:"An optional prefix for emitted New Relic events"`
+		AccountID          string `long:"newrelic-account-id" description:"New Relic Account ID"`
+		APIKey             string `long:"newrelic-api-key" description:"New Relic Insights API Key"`
+		ServicePrefix      string `long:"newrelic-service-prefix" default:"" description:"An optional prefix for emitted New Relic events"`
+		CompressionEnabled bool   `long:"newrelic-metric-compression" description:"New Relic payload compression flag"`
+		FlushInterval      int    `long:"newrelic-flush-interval" default:"60" description:"New Relic metric flush interval in seconds"`
 	}
 
 	singlePayload map[string]interface{}
 	fullPayload   []singlePayload
+	batchPayload  []byte
+
+	BatchEmitter interface {
+		Emit(payload singlePayload) *loggableError
+		Flush() *loggableError
+	}
+
+	batchEmitter struct {
+		client        *http.Client
+		url           string
+		apikey        string
+		payloadQueue  batchPayload
+		lastEmitTime  time.Time
+		compression   bool
+		dataLock      *sync.Mutex
```

Comment:
do we necessarily need locks on this?

e.g., the `influxdb` emitter is able to do well without as the places where one would usually use locks to synchronize access happen serially, thus, requiring no need for locks 🤔